src/Development/IDE/Core/Shake.hs

```diff
@@ -258,7 +258,22 @@ setValues :: IdeRule k v
           -> IO ()
 setValues state key file val = modifyVar_ state $ \vals -> do
     -- Force to make sure the old HashMap is not retained
-    evaluate $ HMap.insert (file, Key key) (fmap toDyn val) vals
+    let v = fmap toDyn val
+    _ <- evaluate v
+    let k = (file, Key key)
+    res <- evaluate $ HMap.insert k v vals
+
+    -- This odd construction is required to prevent a space leak.
+    -- We insert a fully evaluated value, but, somehow, the HashMap
+    -- seems to keep track of the old value. I don't know how.
+    -- Is there a space-leak in unordered-containers?
+    -- The evaluate below fixes it, at the cost of walking the
+    -- HashMap a second time.
+    -- Without this evaluate, leaks of 30Mb/min have been observed.
+    case HMap.lookup k res of
+        Nothing -> pure ()
+        Just v -> void $ evaluate v
+    return res
```

Comment:
I think the problem is that `modifyVar_` is not strict, it doesn't force the updated value. I expect that switching to a strict variant will allow you to get rid of all the uses of `evaluate`s in this function.

Reply:
But we already called `evaluate` on the HashMap after inserting and it’s a strict HashMap. Shouldn’t that be sufficient?

Reply:
My point is that you wouldn't need that call to `evaluate`. By always using a strict `modifyVar_`, you wouldn't have to remember to `evaluate` all the things you modify. Which is error-prone and ugly (IMO). On the other hand, you must remember to always use a strict `modifyVar_` :slightly_smiling_face:

Reply:
But we _did_ call evaluate. So why was the old code leaking space?

Reply:
True, I can't explain that either.